### PR TITLE
Added qxmpp/1.4.0

### PR DESCRIPTION
**Pull request by COM8**
Specify library name and version:  **qxmpp/1.4.0**

[QXmpp](https://github.com/qxmpp-project/qxmpp) is a cross-platform C++ XMPP client and server library. It is written in C++ and uses Qt framework.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/qxmpp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    def requirements(self):
+        self.requires("qt/6.1.2")
+        if self.options.withGstreamer:
+            self.requires("gstreamer/1.19.1")
+            self.requires("glib/2.68.3")
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:33]:
```suggestion

    def configure(self):
        if self.options.shared:
            del self.options.fPIC
            
```

File: recipes/qxmpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = glob.glob("qxmpp-*")[0]
+        rename(extracted_dir, self.name)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:37]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self.name)
```

File: recipes/qxmpp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        patches = self.conan_data["patches"][self.version]
+        for patch in patches:
+            tools.patch(**patch)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:40]:
```suggestion
```

File: recipes/qxmpp/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __get_option_str(self, b: bool) -> str:
+        if b:
+            return "ON"
+        else:
+            return "OFF"
+
+    def build(self):
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:48]:
```suggestion
    def build(self):
        for patch in self.conan_data["patches"][self.version]:
            tools.patch(**patch)
```

File: recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
++
++conan_cmake_configure(REQUIRES qt/6.1.2
++                      GENERATORS cmake_find_package)
++
++conan_cmake_autodetect(settings)
++
++conan_cmake_install(PATH_OR_REFERENCE .
++                    BUILD missing
++                    REMOTE conan-center
++                    SETTINGS ${settings})
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch:36]:
```suggestion
+if(NOT EXISTS "${CMAKE_BINARY_DIR}/conan.cmake")
+  message(STATUS "Downloading conan.cmake from https://github.com/conan-io/cmake-conan")
+  file(DOWNLOAD "https://raw.githubusercontent.com/conan-io/cmake-conan/v0.16.1/conan.cmake"
+                "${CMAKE_BINARY_DIR}/conan.cmake"
+                EXPECTED_HASH SHA256=396e16d0f5eabdc6a14afddbcfff62a54a7ee75c6da23f32f7a31bc85db23484
+                TLS_VERIFY ON)
+endif()
+
+include(${CMAKE_BINARY_DIR}/conan.cmake)
+
+conan_cmake_configure(REQUIRES qt/6.1.2
+                      GENERATORS cmake_find_package)
+
+conan_cmake_autodetect(settings)
+
+conan_cmake_install(PATH_OR_REFERENCE .
+                    BUILD missing
+                    REMOTE conan-center
+                    SETTINGS ${settings})
```
all conan center recipes pre-suppose that conan is already installed

File: recipes/qxmpp/all/patches/002-add_gstreamer_1.4.0.patch [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
++++ b/CMakeLists.txt
+@@ -25,6 +25,8 @@ endif()
+ include(${CMAKE_BINARY_DIR}/conan.cmake)
+ 
+ conan_cmake_configure(REQUIRES qt/6.1.2
++                               gstreamer/1.19.1
++                               glib/2.68.3
+                       GENERATORS cmake_find_package)
+ 
+ conan_cmake_autodetect(settings)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/patches/002-add_gstreamer_1.4.0.patch:13]:
```suggestion
```

File: recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch [hunk 7]
```diff
@@ -0,0 +1,42 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 550ea39a..9ed4d825 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -8,10 +8,34 @@ set(CMAKE_CXX_STANDARD 17)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch:5]:
```suggestion
@@ -8,10 +8,15 @@ set(CMAKE_CXX_STANDARD 17)
```

File: recipes/qxmpp/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+from conans import ConanFile, CMake, tools
+from os import rename
+from os.path import join
+import glob
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:4]:
```suggestion
```

File: recipes/qxmpp/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    license = "LGPL-2.1"
+    homepage = "https://github.com/qxmpp-project/qxmpp"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "Cross-platform C++ XMPP client and server library. It is written in C++ and uses Qt framework."
+    topics = ("qt", "qt6", "xmpp", "xmpp-library", "xmpp-server", "xmpp-client")
+    settings = "os", "compiler", "build_type", "arch"
+    exports_sources = ["patches/*"]
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "withGstreamer": [True, False]}
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:18]:
```suggestion
               "with_gstreamer": [True, False]}
```

File: recipes/qxmpp/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    default_options = {"shared": False,
+                       "fPIC": True,
+                       "withGstreamer": False}
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:21]:
```suggestion
                       "with_gstreamer": False}
```

File: recipes/qxmpp/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    generators = "cmake"
+
+    def requirements(self):
+        self.requires("qt/6.1.2")
+        if self.options.withGstreamer:
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:26]:
```suggestion
        if self.options.with_gstreamer:
```

File: recipes/qxmpp/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            return "OFF"
+
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
+        cmake = CMake(self)
+        cmake.definitions["BUILD_DOCUMENTATION"] = "OFF"
+        cmake.definitions["BUILD_TESTS"] = "OFF"
+        cmake.definitions["BUILD_EXAMPLES"] = "OFF"
+        cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.withGstreamer)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:54]:
```suggestion
        cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.with_gstreamer)
```

File: recipes/qxmpp/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    topics = ("qt", "qt6", "xmpp", "xmpp-library", "xmpp-server", "xmpp-client")
+    settings = "os", "compiler", "build_type", "arch"
+    exports_sources = ["patches/*"]
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "with_gstreamer": [True, False]}
+    default_options = {"shared": False,
+                       "fPIC": True,
+                       "with_gstreamer": False}
+    generators = "cmake"
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:21]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/qxmpp/all/test_package/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+from conans import ConanFile, CMake, tools
+import os
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/test_package/conanfile.py:6]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${PROJECT_SOURCE_DIR}/cmake/modules")
+ 
+-find_package(QT NAMES Qt6 Qt5 REQUIRED COMPONENTS Core Network Xml)
+-find_package(Qt${QT_VERSION_MAJOR} 5.7.0 REQUIRED COMPONENTS Core Network Xml)
++list(APPEND CMAKE_MODULE_PATH ${CMAKE_BINARY_DIR})
++list(APPEND CMAKE_PREFIX_PATH ${CMAKE_BINARY_DIR})
++
++add_definitions("-std=c++11")
++
++
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch:18]:
```suggestion
```

qt6 needs c++17

File: recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch [hunk 16]
```diff
@@ -0,0 +1,23 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 550ea39a..9ed4d825 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -8,10 +8,15 @@ set(CMAKE_CXX_STANDARD 17)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/patches/001-add_conan_1.4.0.patch:5]:
```suggestion
@@ -8,10 +8,12 @@ set(CMAKE_CXX_STANDARD 17)
```

File: recipes/qxmpp/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        cmake = CMake(self)
+        cmake.definitions["BUILD_DOCUMENTATION"] = "OFF"
+        cmake.definitions["BUILD_TESTS"] = "OFF"
+        cmake.definitions["BUILD_EXAMPLES"] = "OFF"
+        cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.with_gstreamer)
+        cmake.configure(source_folder="qxmpp")
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=join(self.name, "LICENSE.LGPL"))
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:58]:
```suggestion
        self.copy("LICENSE.LGPL", dst="licenses", src=self.name)
```

File: recipes/qxmpp/all/conanfile.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    exports_sources = ["patches/*"]
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "with_gstreamer": [True, False]}
+    default_options = {"shared": False,
+                       "fPIC": True,
+                       "with_gstreamer": False}
+    generators = "cmake", "cmake_find_package_multi"
+
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:22]:
```suggestion
    _cmake = None

```

File: recipes/qxmpp/all/conanfile.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+            
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self.name)
+
+    def __get_option_str(self, b: bool) -> str:
+        if b:
+            return "ON"
+        else:
+            return "OFF"
+
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:45]:
```suggestion

    def _configure_cmake(self):
        if self._cmake:
            return self._cmake
        self._cmake = CMake(self)
        self._cmake.definitions["BUILD_DOCUMENTATION"] = "OFF"
        self._cmake.definitions["BUILD_TESTS"] = "OFF"
        self._cmake.definitions["BUILD_EXAMPLES"] = "OFF"
        self._cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.with_gstreamer)
        self._cmake.configure(source_folder="qxmpp")
        return self._cmake

```

File: recipes/qxmpp/all/conanfile.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
+        cmake = CMake(self)
+        cmake.definitions["BUILD_DOCUMENTATION"] = "OFF"
+        cmake.definitions["BUILD_TESTS"] = "OFF"
+        cmake.definitions["BUILD_EXAMPLES"] = "OFF"
+        cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.with_gstreamer)
+        cmake.configure(source_folder="qxmpp")
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:54]:
```suggestion
        cmake = self._configure_cmake()
```

File: recipes/qxmpp/all/conanfile.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    def package(self):
+        self.copy("LICENSE.LGPL", dst="licenses", src=self.name)
+        self.copy("*.h", dst="include/base", src=join(self.name, "base"))
+        self.copy("*.h", dst="include/client", src=join(self.name, "client"))
+        self.copy("*.h", dst="include/server", src=join(self.name, "server"))
+        self.copy("*qxmpp.lib", dst="lib", keep_path=False)
+        self.copy("*.dll", dst="bin", keep_path=False)
+        self.copy("*.so", dst="lib", keep_path=False)
+        self.copy("*.dylib", dst="lib", keep_path=False)
+        self.copy("*.a", dst="lib", keep_path=False)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:66]:
```suggestion
        cmake = self._configure_cmake()
        cmake.install()
```

File: recipes/qxmpp/all/conanfile.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "with_gstreamer": [True, False]}
+    default_options = {"shared": False,
+                       "fPIC": True,
+                       "with_gstreamer": False}
+    generators = "cmake", "cmake_find_package_multi"
+    _cmake = None
+
+    def requirements(self):
```
Comment (uilianries) [inline recipes/qxmpp/all/conanfile.py:24]:
```suggestion
    @property
    def _source_subfolder(self):
        return "source_subfolder"

    def requirements(self):
```

File: recipes/qxmpp/all/conanfile.py [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+            
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self.name)
```
Comment (uilianries) [inline recipes/qxmpp/all/conanfile.py:39]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
```

File: recipes/qxmpp/all/conanfile.py [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+            return "OFF"
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["BUILD_DOCUMENTATION"] = "OFF"
+        self._cmake.definitions["BUILD_TESTS"] = "OFF"
+        self._cmake.definitions["BUILD_EXAMPLES"] = "OFF"
+        self._cmake.definitions["WITH_GSTREAMER"] = self.__get_option_str(self.options.with_gstreamer)
```
Comment (uilianries) [inline recipes/qxmpp/all/conanfile.py:54]:
```suggestion
        self._cmake.definitions["WITH_GSTREAMER"] = self.options.with_gstreamer
```

Cmake is able to solve True/False

File: recipes/qxmpp/all/conanfile.py [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+        self._cmake.configure(source_folder="qxmpp")
```
Comment (uilianries) [inline recipes/qxmpp/all/conanfile.py:55]:
```suggestion
        self._cmake.configure()
```

Use the cmake file wrapper.

File: recipes/qxmpp/all/test_package/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+find_package(qxmpp)
```
Comment (uilianries) [inline recipes/qxmpp/all/test_package/CMakeLists.txt:7]:
```suggestion
find_package(qxmpp REQUIRED)
```

File: recipes/qxmpp/all/test_package/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,12 @@
...
+
+set(CMAKE_CXX_STANDARD 17)
```
Comment (uilianries) [inline recipes/qxmpp/all/test_package/CMakeLists.txt:9]: Does qxmpp require C++ 17? If yes, you need to validate it first on the recipe. Take it as example: https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/cppitertools/all/conanfile.py#L30

Comment (ericLemanissier) [reply recipes/qxmpp/all/test_package/CMakeLists.txt:9]: it does, because qt6 requires c++17

File: recipes/qxmpp/all/conanfile.py [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+        return self._cmake
+
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE.LGPL", dst="licenses", src=self.name)
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:65]:
```suggestion
        self.copy("LICENSE.LGPL", dst="licenses", src=self._source_subfolder)
```

File: recipes/qxmpp/all/conanfile.py [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+            self.requires("glib/2.68.3")
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:37]:
```suggestion

    def validate(self):
        if self.settings.compiler.get_safe("cppstd"):
            tools.check_min_cppstd(self, 17)

```

File: recipes/qxmpp/all/conandata.yml [hunk 30]
```diff
@@ -0,0 +1,10 @@
+sources:
+  "1.4.0":
+    url: "https://github.com/qxmpp-project/qxmpp/archive/refs/tags/v1.4.0.tar.gz"
+    sha256: "2148162138eaf4b431a6ee94104f87877b85a589da803dff9433c698b4cf4f19"
+patches:
+  "1.4.0":
+    - patch_file: "patches/001-add_conan_1.4.0.patch"
+      base_path: "qxmpp"
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conandata.yml:8]:
```suggestion
      base_path: "source_subfolder"
```

File: recipes/qxmpp/all/conandata.yml [hunk 31]
```diff
@@ -0,0 +1,10 @@
...
+    - patch_file: "patches/002-add_gstreamer_1.4.0.patch"
+      base_path: "qxmpp"
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conandata.yml:10]:
```suggestion
      base_path: "source_subfolder"
```

File: recipes/qxmpp/all/conanfile.py [hunk 32]
```diff
@@ -0,0 +1,79 @@
...
+class QxmppConan(ConanFile):
+    name = "qxmpp"
+    license = "LGPL-2.1"
+    homepage = "https://github.com/qxmpp-project/qxmpp"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "Cross-platform C++ XMPP client and server library. It is written in C++ and uses Qt framework."
+    topics = ("qt", "qt6", "xmpp", "xmpp-library", "xmpp-server", "xmpp-client")
+    settings = "os", "compiler", "build_type", "arch"
+    exports_sources = ["patches/*"]
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:14]:
```suggestion
    exports_sources = ["patches/*", "CMakeLists.txt"]
```

File: recipes/qxmpp/all/conanfile.py [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+    def build(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE.LGPL", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:75]:
```suggestion
        cmake.install()
        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```

File: recipes/qxmpp/all/conanfile.py [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+
+    def package_info(self):
+        self.cpp_info.libs = ["qxmpp"]
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:78]:
```suggestion
        self.cpp_info.libs = ["qxmpp"]
        self.cpp_info.names["cmake_find_package"] = "QXmpp"
        self.cpp_info.names["cmake_find_package_multi"] = "QXmpp"
```

File: recipes/qxmpp/all/conanfile.py [hunk 35]
```diff
@@ -0,0 +1,85 @@
+from conans import ConanFile, CMake, tools
+from os import rename
+from os.path import join
```
Comment (ericLemanissier) [inline recipes/qxmpp/all/conanfile.py:3]:
```suggestion
import os
```

**Comment by COM8** [comment]
Closing this one, since I'm not able to get it running.

**Comment by COM8** [comment]
@ericLemanissier Thanks for your help!

**Pull request closed without merging**